client/logmon/logmon.go

```diff
@@ -0,0 +1,215 @@
+	GID int
+}
+
+type LogMon interface {
+	Start(*LogConfig) error
+	Stop() error
+}
+
+func NewLogMon(logger hclog.Logger) LogMon {
+	return &logmonImpl{
+		logger: logger,
+	}
+}
+
+type logmonImpl struct {
+	logger hclog.Logger
+	tl     *TaskLogger
+}
+
+func (l *logmonImpl) Start(cfg *LogConfig) error {
+	tl, err := NewTaskLogger(cfg, l.logger)
+	if err != nil {
+		return err
+	}
+	l.tl = tl
+	return nil
+}
+
+func (l *logmonImpl) Stop() error {
+	l.tl.Close()
+	return nil
+}
+
+type TaskLogger struct {
+	config *LogConfig
+
+	// rotator for stdout
+	lro *logRotatorWrapper
+
+	// rotator for stderr
+	lre *logRotatorWrapper
+}
+
+func (tl *TaskLogger) Close() {
+	if tl.lro != nil {
+		tl.lro.Close()
+	}
+	if tl.lre != nil {
+		tl.lre.Close()
+	}
+}
+
+func NewTaskLogger(cfg *LogConfig, logger hclog.Logger) (*TaskLogger, error) {
+	tl := &TaskLogger{config: cfg}
+
+	logFileSize := int64(cfg.MaxFileSizeMB * 1024 * 1024)
+	lro, err := logging.NewFileRotator(cfg.LogDir, cfg.StdoutLogFile,
+		cfg.MaxFiles, logFileSize, cfg.UID, cfg.GID, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stdout logfile for %q: %v", cfg.StdoutLogFile, err)
+	}
+
+	wrapperOut, err := newLogRotatorWrapper(cfg.StdoutFifo, logger, lro)
+	if err != nil {
+		return nil, err
+	}
+
+	tl.lro = wrapperOut
+
+	lre, err := logging.NewFileRotator(cfg.LogDir, cfg.StderrLogFile,
+		cfg.MaxFiles, logFileSize, cfg.UID, cfg.GID, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stderr logfile for %q: %v", cfg.StderrLogFile, err)
+	}
+
+	wrapperErr, err := newLogRotatorWrapper(cfg.StderrFifo, logger, lre)
+	if err != nil {
+		return nil, err
+	}
+
+	tl.lre = wrapperErr
+
+	return tl, nil
+
+}
+
+// logRotatorWrapper wraps our log rotator and exposes a pipe that can feed the
+// log rotator data. The processOutWriter should be attached to the process and
+// data will be copied from the reader to the rotator.
+type logRotatorWrapper struct {
+	fifoPath          string
+	processOutReader  io.ReadCloser
+	rotatorWriter     *logging.FileRotator
+	hasFinishedCopied chan struct{}
+	logger            hclog.Logger
+}
+
+// newLogRotatorWrapper takes a rotator and returns a wrapper that has the
+// processOutWriter to attach to the processes stdout or stderr.
+func newLogRotatorWrapper(path string, logger hclog.Logger, rotator *logging.FileRotator) (*logRotatorWrapper, error) {
+	logger.Info("opening fifo", "path", path)
+	f, err := fifo.New(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create fifo for extracting logs: %v", err)
+	}
+
+	wrap := &logRotatorWrapper{
+		fifoPath:          path,
+		processOutReader:  f,
+		rotatorWriter:     rotator,
+		hasFinishedCopied: make(chan struct{}),
+		logger:            logger,
+	}
+	wrap.start()
+	return wrap, nil
+}
+
+// start starts a go-routine that copies from the pipe into the rotator. This is
+// called by the constructor and not the user of the wrapper.
```

Comment:
[nitpick] I'm not a big fan of this pattern as it can make unit testing more complex. It's often useful to be able to create a struct with a new func and either peek at initial state or change some fields before running it.

In general I prefer new funcs without side-effects, but it's not a blocker.

You could add a TaskLogger.Start method that starts the wrappers in goroutines and L71 above could call it if New doesn't return an error.